Changes to the shipping rates and suggestions API

nima-karimi opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes #1395.

This PR implements the suggestion changes offered by @ecgan in #1395. For more context please check that issue.

Based on the code introduced in #1341, the suggestions API returned different values that were not expected by the front-end code. This led to some discrepancies between the data structures defined in the UI and the data returned by the API. @ecgan and I discussed this and we concluded that it would be best if the data structures were the same. Given that the suggestions API is only used by our own UI code, I've made some changes to make the API responses aligned with how the UI expects them.

With this PR, only the `flat_rate` shipping method is accepted by the shipping rates APIs. The `free_shipping` method is no longer accepted and to indicate free shipping, a rate of 0 should be set for the shipping method `flat_rate`.

When a `free_shipping_threshold` value is set on a `flat_rate` method, it means that if the minimum order amount hits that value, the shipping rate will be free regardless of the `rate` value.

Moreover, if a country has only a conditional free shipping rate (with a minimum order amount/free_shipping_threshold) and no other rates available, then the conditional free rate will be ignored and no suggestions will be returned for that country.

To parse and format the shipping suggestions, a new ShippingSuggestionService is added with unit tests.

### Screenshots:

<!--- Optional --->


### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

@ecgan I think it would be best if you could test this PR and give a green light since we've discussed this before.

1. Go over the scenarios mentioned in #1395 and confirm the results are as expected.


### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

>

Commit:
Use a single ShippingRate class for all methods

This merged the two classes previously used to define shipping rates (ShippingRateFlat and ShippingRateFree) together. To indicate a free shipping method, we now only set its rate to 0. Moreover, `free_shipping_threshold` option is now renamed to `mini_order_amount` and applies to all shipping rates. This is to be in line with how Google represents the shipping rates.

Commit:
Add shipping suggestions service

This service is used for processing the location rates and returning shipping suggestions for the given country code. It will be used in the shipping suggestions API.

Commit:
Add unit tests for ShippingSuggestionService

Commit:
Use ShippingSuggestionService in the suggestions API controller

jorgemd24 commented on src/Shipping/LocationRatesProcessor.php:
```diff
@@ -56,21 +76,8 @@ public function process( array $location_rates ): array {
-
-		return $replace;
+		return $new_rate->get_rate() > $existing_rate->get_rate() ||
+			   (float) $new_rate->get_min_order_amount() > (float) $existing_rate->get_min_order_amount();
```
❓

I think this should be:

```
		return $new_rate->get_rate() < $existing_rate->get_rate() ||
			   (float) $new_rate->get_min_order_amount() < (float) $existing_rate->get_min_order_amount();
```

Otherwise it will return the most expensive rate.

nima-karimi replied:
This is intentional and specified in the requirements of this feature. The reasoning is that since we are grouping and estimating the rates it would be safer just to assume the most expensive rate for the group rather than applying the cheapest rate. For example, if a country-wide rate is $50 but a state-wide rate is $20, it would make more sense to apply the $50 rate to the whole country rather than the $20 rate.

jorgemd24 replied:
Thanks, @nima-karimi for the clarification, I got confused because [in this comment](https://github.com/woocommerce/google-listings-and-ads/issues/1395#issue-1190146411) we mentioned we would expect to return the cheapest shipping rate:

>Thailand (flat 25, free shipping 0)
Actual: {country: "TH", method: "free_shipping", currency: "USD", rate: 0, options: {}}
Expected (assuming **we are returning cheapest shipping rate**, hence discarding the "flat 25"): {country: "TH", method: "flat_rate", currency: "USD", rate: 0, options: {}}
🔴 Not match - should convert it to flat_rate.

I thought that we would follow the same idea, to suggest the cheapest and not the most expensive. 

On the other hand, I see the point that it is safer to show the most expensive although maybe it is not the best suggestion.

nima-karimi replied:
Oh yeah, that's only the case if there is a free shipping method involved. Although I see the contradictions here. Given the same reasoning, one could argue that we should not return the free shipping but the most expensive flat-rate shipping for that group.

I think it would work best if we return the most expensive rate just to be on the safer side and don't surprise the user with a low rate on their Google profile. I can remove the condition that sets all rates to free if there is a free_shipping rate involved. @jorgemd24 @ecgan Do you think this would be preferable? Using the same Thailand example shared above, we would return the $25 rate instead of the $0

jorgemd24 replied:
In my opinion, the issue is because it is difficult to make the balance between "safe" and a good suggestion only based on the rates that are set up in WC.

It sounds right that for example if Thailand has free shipping it should return a rate of $0 although, as you mentioned before, it is contradictory to the requirements of this feature. If we think from the "safer side", the free shipping could never apply if there is other flat rate > $0.

When there is more than one flat rate > $0, it is impossible to know which one is the right one so I suppose it is good to go with the most expensive, although this can go against the user's expectations. 

We need just to make sure that everyone understands that we are not returning the cheapest shipping rate, we return free shipping if exists otherwise we return the most expensive.

nima-karimi replied:
Yes, it's a trade-off. And we're merely guessing the rates to make the onboarding smoother and also telling them to always check them in their Merchant Center settings after the onboarding is done. I think to be consistent it would be best to always return the most expensive rate regardless of whether they have a free-shipping rate or not. I've made some changes in https://github.com/woocommerce/google-listings-and-ads/pull/1422/commits/4f79a601e47dd46fbb9634a4b9297269214414af to always return the most expensive rate.

jorgemd24 commented on src/Shipping/ShippingSuggestionService.php:
```diff
@@ -0,0 +1,96 @@
+				$free_threshold = $shipping_rate->get_min_order_amount();
+
+				// Ignore the conditional free rate if there are other rates.
+				if ( count( $location_rates ) > 1 ) {
```
❔

I am not really sure why this is necessary, for example if we have a free shipping with a min. order amount and no other rates, the rate is skipped anyway. 

I think we could remove `if ( count( $location_rates ) > 1 )` but I do not know if I am missing something?

nima-karimi replied:
> I am not really sure why this is necessary, for example if we have a free shipping with a min. order amount and no other rates, the rate is skipped anyway.

This is for the inverse condition, meaning that we have conditional free shipping rates AND other rates. What this part does is that it gets the `min_order_amount` value from a conditional free shipping rate and adds it as a `free_shipping_threshold` to all the other rates. When we add that value to the other rates, we need to remove the free shipping rate from the list because it's already implied by that `free_shipping_threshold` value in other rates.

This is just a data format change so that it can be returned the way UI expects it.

Instead of returning this data:

```JSON
    {
        "country": "US",
        "method": "flat_rate",
        "currency": "USD",
        "rate": 20
    }
    {
        "country": "US",
        "method": "flat_rate",
        "currency": "USD",
        "rate": 0,
        "min_order_amount": 100
    }
```

We would return the following in the API response:

```JSON
    {
        "country": "US",
        "method": "flat_rate",
        "currency": "USD",
        "rate": 20,
        "options": {
            "free_shipping_threshold": 100
        }
    }
```

nima-karimi replied:
> I think we could remove if ( count( $location_rates ) > 1 ) but I do not know if I am missing something?

I rather keep this condition because if we remove it then the code here would depend on the conditions defined in other classes with no way to check if the logic is as expected. This condition helps make sure the conditional free rate is always ignored when its parameters are applied to the other rates.

jorgemd24 replied:
Thanks @nima-karimi that makes sense 👍

jorgemd24 commented on tests/Unit/Shipping/LocationRatesProcessorTest.php:
```diff
@@ -22,69 +21,88 @@ class LocationRatesProcessorTest extends UnitTest {
-		$flat_rate = $processed[0]->get_shipping_rate();
-		$this->assertInstanceOf( ShippingRateFlat::class, $flat_rate );
-		$this->assertEquals( 200, $flat_rate->get_rate() );
+		$this->assertEquals( 200, $processed[0]->get_shipping_rate()->get_rate() );
```
❓

Should not return the cheapest value? 50?

nima-karimi replied:
No it's intentionally returning the most expensive rate as I mentioned in https://github.com/woocommerce/google-listings-and-ads/pull/1422#discussion_r851109599

jorgemd24 commented on tests/Unit/Shipping/LocationRatesProcessorTest.php:
```diff
@@ -22,69 +21,88 @@ class LocationRatesProcessorTest extends UnitTest {
+		$this->assertEquals( 2, count( $processed ) );
+		foreach ( $processed as $location_rate ) {
+			if ( $location_rate->get_shipping_rate()->is_free() ) {
+				$this->assertEquals( 200, $location_rate->get_shipping_rate()->get_min_order_amount() );
```
❓

Should not return 50 as the minum order amount?

nima-karimi replied:
see https://github.com/woocommerce/google-listings-and-ads/pull/1422#discussion_r851109599

The same rule applies to the minimum order amount because we wouldn't want to set a low minimum order amount for a group if it only applies to a state or postcode within a country, etc.

jorgemd24 reviewed (commented):
Thanks, @nima-karimi 👍  I have left some comments about the code, in general, is working as expected, except that is returning the most expensive rate instead of the cheapest.

ecgan commented:
@nima-karimi ,

> @ecgan I think it would be best if you could test this PR and give a green light since we've discussed this before.
> 
> 1. Go over the scenarios mentioned in [Better shipping rates suggestions #1395](https://github.com/woocommerce/google-listings-and-ads/issues/1395) and confirm the results are as expected.

I tested this out using this WC shipping zones setting as per #1395, and I can confirm this works perfect and matches the expectations in #1395! 💯  

When I visit the Setup MC shipping rate section, and selects the "Simple" flat shipping rate options, this is the UI that I get, which is all good 👍 :

![image](https://user-images.githubusercontent.com/417342/163450280-53089e93-3510-47d8-92e9-61d27a077137.png)

This is the shipping rate suggestions API response data that I got:

```json
[
  {
    "country": "ID",
    "method": "flat_rate",
    "currency": "USD",
    "rate": 0,
    "options": {}
  },
  {
    "country": "MM",
    "method": "flat_rate",
    "currency": "USD",
    "rate": 25,
    "options": {}
  },
  {
    "country": "MY",
    "method": "flat_rate",
    "currency": "USD",
    "rate": 0,
    "options": {}
  },
  {
    "country": "PH",
    "method": "flat_rate",
    "currency": "USD",
    "rate": 0,
    "options": {}
  },
  {
    "country": "SG",
    "method": "flat_rate",
    "currency": "USD",
    "rate": 0,
    "options": {}
  },
  {
    "country": "TH",
    "method": "flat_rate",
    "currency": "USD",
    "rate": 0,
    "options": {}
  },
  {
    "country": "LK",
    "method": "flat_rate",
    "currency": "USD",
    "rate": 25,
    "options": { "free_shipping_threshold": 50 }
  }
]

```

---

One side question: 

> With this PR, only the `flat_rate` shipping method is accepted by the shipping rates APIs. The `free_shipping` method is no longer accepted and to indicate free shipping, a rate of 0 should be set for the shipping method `flat_rate`.

Will there be any method other than `flat_rate` in the future? If there is only going to be one `flat_rate` method, maybe we can consider removing `method` field from shipping rates in future PR? 🤔  

Note: if we were to remove the `method` field, UI front-end code change is required.

nima-karimi commented:
> Will there be any method other than flat_rate in the future? If there is only going to be one flat_rate method, maybe we can consider removing method field from shipping rates in future PR? 🤔

I thought about doing this but we do have the pick up method that might be introduced in the front-end. I'm not sure if we are ever going to introduce options for the user to set a `pickup` method in the UI. If we don't have any intentions of doing that in the near future then we can remove it completely.

Commit:
Remove shipping `method`

Shipping method is no longer relevant, and we can add it back later when we have more than one method available

Commit:
Fix code style

Commit:
Set default type to `flat_rate` when processing rates

Commit:
Return most expensive shipping rate suggestion

Even if there is an unconditional free shipping in the group.

nima-karimi commented:
@ecgan I've removed the `method` from all shipping rate-related APIs and also the DB tables. You can create another PR off of this one to apply the necessary changes to the UI code.

I don't think we'll be adding the pickup method to the UI anytime soon since it clutters the current UI and goes against the phrase "My shipping settings are simple".

nima-karimi commented:
@jorgemd24 @ecgan Thanks for all the comments. I think it would be best to do another round of testing given that I've changed some logic in the last commits to this PR. To test the UI, @ecgan needs to first apply the changes.

ecgan commented:
@nima-karimi , 

> @ecgan I've removed the `method` from all shipping rate-related APIs and also the DB tables. You can create another PR off of this one to apply the necessary changes to the UI code.

Okay, I'll work on that.

> I think it would be best to do another round of testing given that I've changed some logic in the last commits to this PR.

I tested the latest code change here with the same WC shipping zones settings. Below is the suggestions API response:

```json
[
    {
        "country": "ID",
        "currency": "USD",
        "rate": 0,
        "options": {}
    },
    {
        "country": "MM",
        "currency": "USD",
        "rate": 25,
        "options": {}
    },
    {
        "country": "MY",
        "currency": "USD",
        "rate": 0,
        "options": {}
    },
    {
        "country": "PH",
        "currency": "USD",
        "rate": 0,
        "options": {}
    },
    {
        "country": "SG",
        "currency": "USD",
        "rate": 0,
        "options": {
            "free_shipping_threshold": 50
        }
    },
    {
        "country": "TH",
        "currency": "USD",
        "rate": 25,
        "options": {}
    },
    {
        "country": "LK",
        "currency": "USD",
        "rate": 25,
        "options": {
            "free_shipping_threshold": 50
        }
    }
]
```

Other than the removal of `method`, there are two countries that are different from the previous API response in https://github.com/woocommerce/google-listings-and-ads/pull/1422#issuecomment-1099495644:

- For `SG`, we have rate 0 and free_shipping_threshold 50. That doesn't look right? 🤔 
- For `TH`, the rate is now 25 instead of the previous 0. This is expected since we are now suggesting the most expensive rate, as per https://github.com/woocommerce/google-listings-and-ads/pull/1422#discussion_r851197518.

Commit:
Ignore conditional free rates if there is another unconditional one.

nima-karimi commented:
Thanks a lot for testing this again, @ecgan 🙏 I've made some changes in https://github.com/woocommerce/google-listings-and-ads/pull/1422/commits/10a74e75f34d22cfe242867d97d0b899eef23af8 that should resolve the issue with the `SG` case. Let me know how it works out.

ecgan commented:
@nima-karimi , thank you. I checked out your code and here's the suggestions API response:

```json
[
    {
        "country": "ID",
        "currency": "USD",
        "rate": 0,
        "options": {}
    },
    {
        "country": "MM",
        "currency": "USD",
        "rate": 25,
        "options": {}
    },
    {
        "country": "MY",
        "currency": "USD",
        "rate": 0,
        "options": {}
    },
    {
        "country": "PH",
        "currency": "USD",
        "rate": 0,
        "options": {}
    },
    {
        "country": "SG",
        "currency": "USD",
        "rate": 0,
        "options": {}
    },
    {
        "country": "TH",
        "currency": "USD",
        "rate": 25,
        "options": {}
    },
    {
        "country": "LK",
        "currency": "USD",
        "rate": 25,
        "options": {
            "free_shipping_threshold": 50
        }
    }
]
```

For the "SG" case, now it has rate 0 without the free_shipping_threshold option. Seems all good now. 💯 🙂

jorgemd24 reviewed (commented):
Thanks @nima-karimi for the changes, I had a look at the Google MC and I saw that the name of the rate was "false" and it is because the following warning:  `WARNING sprintf(): Too few arguments on line number 10`

![image](https://user-images.githubusercontent.com/2488994/163977798-a2ef9cb9-faae-4f33-a2ca-63ecec41d65f.png)

https://github.com/woocommerce/google-listings-and-ads/blob/10a74e75f34d22cfe242867d97d0b899eef23af8/src/API/Google/Settings.php#L300-L305

It should be ` 'Flat rate - %1$s %2$s'`:

I also notice that when there is a country for example Italy with a flat rate: of $5 and  free shipping with a minimum order is not syncing properly.

I get two flat rates in the MC - One flat rate of $5 and a flat rate of $0 and I can't see the minimum order restriction.

![image](https://user-images.githubusercontent.com/2488994/163979252-1c03ed9a-3667-41c3-b1fb-e9437215528a.png)

![image](https://user-images.githubusercontent.com/2488994/163979844-1c35a13b-b861-42f7-9cc9-c46cce8bdc7c.png)


![image](https://user-images.githubusercontent.com/2488994/163979480-15e1ccd7-7a03-4dca-a758-1b5afc8d1bb5.png)

Commit:
Fix shipping name string format

nima-karimi commented:
Thanks, @jorgemd24 

> It should be 'Flat rate - %1$s %2$s':

I fixed this in https://github.com/woocommerce/google-listings-and-ads/pull/1422/commits/720c0c00b2a4826c5363a821919419af66b0b1f3

> I also notice that when there is a country for example Italy with a flat rate: of $5 and free shipping with a minimum order is not syncing properly.

Does this happen when you choose the "Recommended" option or the manual option to sync the settings? Can you try checking out the branch for the https://github.com/woocommerce/google-listings-and-ads/pull/1455 PR and see if this works there? I've made some changes in that PR that will essentially overwrite a bunch of codes in here once it's merged.

jorgemd24 commented:
Thanks @nima-karimi, 

>Does this happen when you choose the "Recommended" option or the manual option to sync the settings? 

Yeah, I chose the "Recommended" option.

>Can you try checking out the branch for the https://github.com/woocommerce/google-listings-and-ads/pull/1455 PR and see if this works there? 

I have tried this scenario:

Spain Flat Rate: $15
Italy: Flat Rate of $5 and free shipping minimum order of $10.

I got the following in the MC:

https://user-images.githubusercontent.com/2488994/166671873-ab7a39b1-640f-48eb-83e0-6304cc92331c.mov

- I see three rates and all of them appear with the name "All Products" and the Shipping rate name as empty, is this intended? 
- The rate for Spain is OK
- The rate for Italy shows two rates one of $5 and the other one of $0, with a minimum order of $10. So it looks good.

Let's wait until the other PR is reviewed so we can check that everything is working fine here.

nima-karimi commented:
> I see three rates and all of them appear with the name "All Products" and the Shipping rate name as empty, is this intended?

Yeah, I intentionally removed the rate name because it was optional and I didn't see any benefits to adding it.

Commit:
Remove ShippingZone dependency from controller

nima-karimi commented:
@jorgemd24 👋 https://github.com/woocommerce/google-listings-and-ads/pull/1455 is now approved and ready to merge. I rather get this PR approved first before merging the other one because it's a rather large PR and it will further complicate the review of this one 😅 Let me know what you think. Would it be okay for you if I merge https://github.com/woocommerce/google-listings-and-ads/pull/1455 into this one so you can test the functionality?

jorgemd24 reviewed (approved):
Hey @nima-karimi!

> I rather get this PR approved first before merging the other one because it's a rather large PR and it will further complicate the review of this one 😅

Yeah, you are right, anyway I tested that is returning the correct shipping rates for different methods, and I can confirm that it is returning the most expensive rate and the property method has been removed.

The problems syncing to the MC are not relevant anymore as the method name is not used (before it was syncing as `false`) and using PR #1455 I see that is syncing the rates properly.
